### PR TITLE
docs: New docs for table functions

### PR DESCRIPTION
**Pull request by purplefox**
### Description

Fixes https://github.com/confluentinc/ksql/issues/3509 https://github.com/confluentinc/ksql/issues/3510

This PR is stacked upon https://github.com/confluentinc/ksql/pull/3702 - please do not review the previous commits.

This PR contains the docs for the new table function functionality. It includes:

* New syntax reference for built in table functions
* New syntax reference for new UDFs
* New section in dev guide on table functions

### Testing done 

N/A

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/table-functions.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+
+Using Table Functions With KSQL
+###############################
+
+A _table function_ is a function that returns a set of zero or more rows. Contrast this to a scalar
+function which returns a single value.
+
+Table functions can be thought of as analagous to the `FlatMap` operation commonly found in
+functional programming or stream processing frameworks such as Kafka Streams.
```
Comment (JimGalasyn) [inline docs/developer-guide/table-functions.rst]:
```suggestion
functional programming or stream processing frameworks such as |kstreams|.
```

File: docs/developer-guide/table-functions.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+Table functions are used in the SELECT clause of a query. They cause the query to potentially
+output more than one row for each input value.
+
+The current implementation of table functions only allows a single column to be returned. This column
+can be any valid KSQL type.
+
+Here's an example of the `EXPLODE` built-in table function which takes an ARRAY and outputs one value
```
Comment (JimGalasyn) [inline docs/developer-guide/table-functions.rst]:
```suggestion
Here's an example of the ``EXPLODE`` built-in table function which takes an ARRAY and outputs one value
```

File: docs/developer-guide/table-functions.rst [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  {sensor_id:54321 reading: 65}
+  {sensor_id:54321 reading: 38}
+
+When scalar values are mixed with table function returns in a SELECT clause, the scalar values
+(in the above example `sensor_id`) are copied for each value returned from the table function.
+
+You can also use multiple table functions in a SELECT clause. In this situation, the results of the
+table functions are `zipped` together. The total number of rows returned is equal to the greatest
+number of values returned from any of the table functions. If some of the functions return fewer
+rows than others, the missing values are replaced with `null`.
```
Comment (JimGalasyn) [inline docs/developer-guide/table-functions.rst]:
```suggestion
rows than others, the missing values are replaced with ``null``.
```

File: docs/developer-guide/table-functions.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  {country: 'UK', name: 'george', age: 3}
+  {country: 'UK', name: 'ringo', age: null}
+  {country: 'USA', name: 'chad', age: 56}
+  {country: 'USA', name: 'chip', age: 34}
+  {country: 'USA', name: 'brad', age: 76}
+  {country: 'USA', name: null, age: 84}
+  {country: 'USA', name: null, age: 56}
+
+
+Built-on Table Functions
```
Comment (JimGalasyn) [inline docs/developer-guide/table-functions.rst]:
```suggestion
Built-in Table Functions
```

File: docs/developer-guide/table-functions.rst [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+========================
+
+KSQL comes with built-in table functions. Please see the syntax reference for more
```
Comment (JimGalasyn) [inline docs/developer-guide/table-functions.rst]:
```suggestion
KSQL comes with built-in table functions. For more information, see :ref:`ksql_table_functions`.
```

File: docs/developer-guide/table-functions.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
+.. _table-functions:
+
+
+Using Table Functions With KSQL
+###############################
+
+A _table function_ is a function that returns a set of zero or more rows. Contrast this to a scalar
+function which returns a single value.
+
+Table functions can be thought of as analagous to the `FlatMap` operation commonly found in
```
[the remark this answers is not in the document]
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Hey JIm, this PR is a WIP (hence the comment in the description). Could you hold off from reviewing for a bit? :)

[the remark this answers is not in the document]
Comment (JimGalasyn) [reply docs/developer-guide/table-functions.rst]: No problemo!

[the remark this answers is not in the document]
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst (hunk 1 above)
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst (hunk 2 above)
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  {sensor_id:12345 reading: 56}
+  {sensor_id:12345 reading: 3}
+  {sensor_id:12345 reading: 76}
+  {sensor_id:12345 reading: 75}
+  {sensor_id:54321 reading: 12}
+  {sensor_id:54321 reading: 65}
+  {sensor_id:54321 reading: 38}
+
+When scalar values are mixed with table function returns in a SELECT clause, the scalar values
+(in the above example `sensor_id`) are copied for each value returned from the table function.
```
[the remark this answers is not in the document]
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+You can also use multiple table functions in a SELECT clause. In this situation, the results of the
+table functions are `zipped` together. The total number of rows returned is equal to the greatest
```
[the remark this answers is not in the document]
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst (hunk 3 above)
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst (hunk 4 above)
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst (hunk 5 above)
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

File: docs/developer-guide/table-functions.rst [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  {country: 'USA', name: 'brad', age: 76}
+  {country: 'USA', name: null, age: 84}
+  {country: 'USA', name: null, age: 56}
+
+
+Built-on Table Functions
+========================
+
+KSQL comes with built-in table functions. Please see the syntax reference for more
+information.
```
[the remark this answers is not in the document]
Comment (purplefox) [reply docs/developer-guide/table-functions.rst]: Fixed

**Comment by purplefox** [comment]
Hi @JimGalasyn I have addressed your previous comments, and the PR now also contains some new syntax for new UDFs ENTRIES and RANGE, any chance you could take a look at those? Thanks

**Pull request closed without merging**